lib/src/source_visitor.dart

```diff
@@ -886,17 +886,35 @@ class SourceVisitor extends ThrowingAstVisitor {
 
     // If the body is a binary operator expression, then we want to force the
     // split at `=>` if the operators split. See visitBinaryExpression().
-    if (node.expression is! BinaryExpression) builder.endRule();
+    var expression = node.expression;
+    if (expression is! BinaryExpression) builder.endRule();
 
     if (_isInLambda(node)) builder.endSpan();
 
-    builder.startBlockArgumentNesting();
+    // If the body is a block-formatted function invocation, don't add extra
+    // nesting to preserve normal indentation for decorators that modify the
+    // functionality of a method body:
+    //
+    //    void foo() => decorate(() {
+    //      "the body";
+    //    });
+    var isDecorator = expression is InvocationExpression &&
```

Comment:
I did some testing locally, and I think using it for lambdas generates too many false positives, so I think we should only support this for named declarations.